### PR TITLE
[10_6_X] DDXv2 and BTV changes in Nano #31618, #31786

### PR DESCRIPTION
**Pull request by andrzejnovak**
Backport of #31618, #31786

- DDX "v1" is kept alongside of v2
- NanoAOD era modifiers are synced to master

### CODE REVIEW

**Comment by mariadalfonso** [comment]
@andrzejnovak 

this backport PR seems take care only of the previously made nano. but not of the remini-UL and nano that come from there.

In master https://github.com/cms-sw/cmssw/pull/31618 you have modifications in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py
and this V2 should be the UL-reminiAOD default.
Then in nano from the UL-remini we do not re-run the V2 and we just get from the mini.

**Comment by andrzejnovak** [comment]
@mariadalfonso Ok so that was not clear to me. The 106 mini should by default use DDX v2? I thought this was not desirable since there already exists 106 mini which does not contain the V2.

**Comment by mariadalfonso** [comment]
@andrzejnovak 
it's correct that the 106 mini out of the box should be unmodified.

That's why we need to use modifiers.
The run2_miniAOD_UL will be the new type of mini when we have the reminiAOD and any modification should be beyond this modifier.
The V2 can be the default once the run2_miniAOD_UL is activated.

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 1]
```diff
@@ -253,6 +255,15 @@
...
+for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016, run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2, run2_nanoAOD_102Xv1, run2_nanoAOD_106Xv1:
+  # Deprecated after 106X
+  modifier.toModify(jetTable.variables,
+    btagCMVA = Var("bDiscriminator('pfCombinedMVAV2BJetTags')",float,doc="CMVA V2 btag discriminator",precision=10),
+    btagDeepC = Var("bDiscriminator('pfDeepCSVJetTags:probc')",float,doc="DeepCSV charm btag discriminator",precision=10),
+    btagDeepFlavC = Var("bDiscriminator('pfDeepFlavourJetTags:probc')",float,doc="DeepFlavour charm tag discriminator",precision=10),
+    )
+for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016:
+  modifier.toModify( jetTable.variables, jetId = Var("userInt('tightIdLepVeto')*4+userInt('tightId')*2+userInt('looseId')",int,doc="Jet ID flags bit1 is loose, bit2 is tight, bit3 is tightLepVeto"))
 run2_jme_2016.toModify( jetTable.variables, jetId = Var("userInt('tightIdLepVeto')*4+userInt('tightId')*2+userInt('looseId')",int,doc="Jet ID flags bit1 is loose, bit2 is tight, bit3 is tightLepVeto"))
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/jets_cff.py:267]: this line seems duplicated/modified and not needed.

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 2]
```diff
@@ -317,6 +324,7 @@ def nanoAOD_customizeCommon(process):
         nanoAOD_addDeepBTag_switch = cms.untracked.bool(False),
         nanoAOD_addDeepBoostedJet_switch = cms.untracked.bool(True),
         nanoAOD_addDeepDoubleX_switch = cms.untracked.bool(True),
+        nanoAOD_addDeepDoubleXV2_switch = cms.untracked.bool(True),
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:327]: few lines below need to reset as well the nanoAOD_addDeepDoubleXV2_switch 

 ```
  run2_miniAOD_80XLegacy.toModify(nanoAOD_addDeepInfoAK8_switch,
                                    nanoAOD_addDeepBTag_switch = True,
                                    nanoAOD_addDeepBoostedJet_switch = False,
                                    nanoAOD_addDeepDoubleX_switch = False,
                                    **nanoAOD_addDeepDoubleXV2_switch = False,**
                                    nanoAOD_addParticleNet_switch = False,
                                    jecPayload = 'AK8PFchs')
```

Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/nano_cff.py:327]: for run2_nanoAOD_106Xv1  something is done differently than in master:
we are re-running the addDeepDoubleX (and addDeepBoostedJet), any reason for not taking directly from miniAOD like in master ?

I think here we miss this block  like this
 run2_nanoAOD_106Xv1.toModify(
       nanoAOD_addDeepInfoAK8_switch,
       nanoAOD_addDeepBoostedJet_switch = False,
       nanoAOD_addDeepDoubleX_switch = False,
    )

**Comment by mariadalfonso** [comment]
@andrzejnovak 
resending in case you missed my comments

**Comment by andrzejnovak** [comment]
Sorry for the delay, hopefully, I understood the requested changes correctly.

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 3]
```diff
@@ -253,7 +255,15 @@
...
 ### Era dependent customization
-run2_jme_2016.toModify( jetTable.variables, jetId = Var("userInt('tightIdLepVeto')*4+userInt('tightId')*2+userInt('looseId')",int,doc="Jet ID flags bit1 is loose, bit2 is tight, bit3 is tightLepVeto"))
+for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016, run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2, run2_nanoAOD_102Xv1, run2_nanoAOD_106Xv1:
+  # Deprecated after 106X
+  modifier.toModify(jetTable.variables,
+    btagCMVA = Var("bDiscriminator('pfCombinedMVAV2BJetTags')",float,doc="CMVA V2 btag discriminator",precision=10),
+    btagDeepC = Var("bDiscriminator('pfDeepCSVJetTags:probc')",float,doc="DeepCSV charm btag discriminator",precision=10),
+    btagDeepFlavC = Var("bDiscriminator('pfDeepFlavourJetTags:probc')",float,doc="DeepFlavour charm tag discriminator",precision=10),
+    )
+for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016, run2_jme_2016:
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/jets_cff.py:265]: remove the run2_jme_2016 from this list

there is no need to modify this tightIdLepVeto see master
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/jets_cff.py#L271-L274

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 4]
```diff
@@ -317,27 +324,33 @@ def nanoAOD_customizeCommon(process):
...
-        )
+    # Don't rerun where already present
+    for modifier in run2_miniAOD_devel, run2_nanoAOD_106Xv1
+        modifier.toModify(
+            nanoAOD_addDeepInfoAK8_switch,
+            nanoAOD_addDeepBoostedJet_switch = False,
+            nanoAOD_addDeepDoubleX_switch = False,
+            nanoAOD_addDeepDoubleXV2_switch = False,
+            nanoAOD_addParticleNet_switch = False,
+            )
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:348]: This is ok for the run2_miniAOD_devel that is going to serve the UL-reminiAOD
but is not ok for the run2_nanoAOD_106Xv1 (already made UL miniAOD) where the ParticleNet and DeepDoubleXV2 do not exist yet.

easiest is to add a separate block for the two things needed

   for modifier in run2_miniAOD_devel, run2_nanoAOD_106Xv1
         modifier.toModify(
             nanoAOD_addDeepInfoAK8_switch,
             nanoAOD_addDeepBoostedJet_switch = False,
             nanoAOD_addDeepDoubleX_switch = False,
             )

**Comment by mariadalfonso** [comment]
please test

**Comment by slava77** [comment]
> Backport of #31618, #31786

this doesn't look like a complete description
- the old DDX "v1" (not versioned) is kept (it is dropped in the master)
    - was this the intention to keep "v1"? If so, please update the PR description.
- in PhysicsTools/NanoAOD/python/jets_cff.py I see what looks like a contribution from #31303 
    - I guess this is required, but seems unrelated to DDXv2

**Comment by mariadalfonso** [comment]
@slava77 
@andrzejnovak

> * in PhysicsTools/NanoAOD/python/jets_cff.py I see what looks like a contribution from #31303
>   
>   * I guess this is required, but seems unrelated to DDXv2

this is actually not needed in 10_6 because we didn't to the cleanup as in https://github.com/cms-sw/cmssw/pull/31213

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 5]
```diff
@@ -253,7 +255,15 @@
 jetTable.variables.pt.precision=10
 
 ### Era dependent customization
-run2_jme_2016.toModify( jetTable.variables, jetId = Var("userInt('tightIdLepVeto')*4+userInt('tightId')*2+userInt('looseId')",int,doc="Jet ID flags bit1 is loose, bit2 is tight, bit3 is tightLepVeto"))
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/jets_cff.py:256]: this PR as now at commit 39aa5df
change the behaviour of the 2016 jetID for the run2_nanoAOD_106Xv1, I can see also from the DQM plot.
restoring the behaviour of master.

Comment (andrzejnovak) [reply PhysicsTools/NanoAOD/python/jets_cff.py:256]: @mariadalfonso no changes are required then?

Comment (ahinzmann) [reply PhysicsTools/NanoAOD/python/jets_cff.py:256]: looks like this line is wrong in 10_6_X.
It should be modified like in the master:
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/jets_cff.py#L271
for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016:
  modifier.toModify( jetTable.variables, jetId = Var("userInt('tightIdLepVeto')*4+userInt('tightId')*2+userInt('looseId')",int,doc="Jet ID flags bit1 is loose, bit2 is tight, bit3 is tightLepVeto"))
Would you mind fixing this here?

Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/jets_cff.py:256]: @ahinzmann thanks for the quick feedback
the fix is done with this PR at commit 39aa5df restore the behaviour of master for this

@andrzejnovak
please add in the PR description a comment on this, together with Slava's requests

**Comment by slava77** [comment]
@cmsbuild please test

@smuzaffar @mrodozov 
it looks like the baseline IB outputs were cleaned up all the way through CMSSW_10_6_X_2020-11-01-0000 . Is the retention minimum now down to less than 2 days? Please check, I recall that the goal was to have it for a week minimum.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31847	39aa5df 
- concerning reco package changes only: the backport of #31618 is appropriate, even though it's not direct; DDXv2 taggers were added for `run2_miniAOD_UL` rather than replacing the v1 as was done in the master (this is now provided in the updated PR description from a few days ago).
    - it looks like the PhysicsTools/NanoAOD part is still under discussion
- jenkins tests pass and comparisons with the baseline show differences in 
    - the UL re-miniAOD workflow 136.88811: 6 new tags were added in slimmedJetsAK8, as expected
    - nanoAOD plots have changes both in the "from scratch" case in wf 10824.0 and in nanoAOD wf made from 2017 MC inputs in wf 1325.7. It sounds like these are still under review by xpog.

**Comment by mariadalfonso** [comment]
+xpog

Changes inline with master:
* Fat Jet have the new DeepTagV2 for all eras
* AK4Jets have btagDeepFlav redefined as in master

Additionally:

- [ ] for the 2016MC of UL type we have included a fix for the jetID that missed to be backported as discussed here https://github.com/cms-sw/cmssw/pull/31847#discussion_r514154284

- [ ] there are two spurious lines that have no effect related to a cleanup that happened in master but not backported
https://github.com/cms-sw/cmssw/pull/31847/files#diff-377770d6a7d4705da48aa96709bdedefcb42d5b5497c85ba41a3f11074c148a0R480-R481
cleanup can be done after

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**